### PR TITLE
refactor(plugins-iterator): refactor precedence matching algoritm

### PR DESCRIPTION
**Pull request by jschmid1**
### Data structure to store plugin scope configurations

This patch set changes how the Plugins Iterator finds suitable plugin configurations for defined precedence rules.

To achieve that, we move from a nested table structure that stores various combinations of `route, service and consumer` to a _flat_ hashmap.

**Before.**

```lua
combos[name].rsc[cfg.route_id] =
combos[name].rsc[cfg.route_id] or {}
combos[name].rsc[cfg.route_id][cfg.service_id] =
combos[name].rsc[cfg.route_id][cfg.service_id] or {}
combos[name].rsc[cfg.route_id][cfg.service_id][cfg.consumer_id] = cfg

combos[name].rc[cfg.route_id] =
combos[name].rc[cfg.route_id] or {}
combos[name].rc[cfg.route_id][cfg.consumer_id] = cfg
-- and so on for all possible combinations
```

**After:**

```lua
-- Build a compound key using the route_id, service_id, and consumer_id
local compound_key = PluginsIterator.build_compound_key(route_id, service_id, consumer_id)

-- Store the plugin configuration in the 'combos' table using the compound key as index
combos[name][compound_key] = cfg
```

The `build_compound_key` function concatenates the `ids` of scoped entity to generate a unique index.

### Lookup algorithm

The change to the data structure allows us to move away from the concept of [combo_keys](https://github.com/Kong/kong/blob/fb6aeff909a508cca6746acd5711cea5d927b6d5/kong/runloop/plugins_iterator.lua#L24-L31) to an index based lookup.

**Before:**

```lua
if route_id and service_id and consumer_id and combos[COMBO_RSC]
  and combos.rsc[route_id] and combos.rsc[route_id][service_id]
  and combos.rsc[route_id][service_id][consumer_id]
then
  return combos.rsc[route_id][service_id][consumer_id]
end
```

**After:**

```lua
return combos[build_compound_key(route_id, service_id, consumer_id)] or
	-- ...
```

### Aligning db-less and db modes

This patch also streamlines the way these different modes handle configuration retrieval.

In traditional mode, we stored a placeholder value of `true` in the lookup tables to indicate that a configuration is in fact present. This would trigger a cache retrieval with transparent database lookup in case of a cache-miss.

In db-less mode, we saved the entire configuration alongside the lookup table. This reduces the lookup-time by avoiding potentially costly detours to the database, as well as the extra function call to `cache`

The obvious drawbacks are an increased memory footprint _for traditional_ _mode_ as we're saving configuration data alongside the lookup table. 
In _db-less_ mode, the memory footprint should be slightly reduced due to a more efficient/predictable table structure.

This change also allows to drop cache-warmups during Plugins Iterator rebuilds, which speeds rebuild times in _traditional mode_.

### Tests

Tests to support the validity of this change are here:

`kong/spec/02-integration/15-plugins-iterator`

### Performance measurements

Where `baseline` is a nightly image without these changes
and `variant` is containing the described changes

Test scenario:
10.000 configured plugins scoped to separate routes, ramping pressure with 10k req/s peak load

- Traditional mode

image

- Hybrid mode

image

##### Interpreting the measurements

Overall, the increases or decreases in performance are not significant enough to claim a real impact.

### CODE REVIEW

**Comment by bungle** [comment]
> Aligning db-less and db modes

@hbagdi, I think this has been proposed before, but we have always refused to do it because of "consumer" entity, and thus we have had many discussions about "getting rid of consumers". The problem comes from the traditional mode use case, where there are unbounded number of consumers (cardinality of consumers can be low or extremely high, and then again there may or may not be plugins attached to each consumer), and then associated plugin (potentially millions). Sure we have the same problem if we try to migrate such traditional customer to hybrid or dbless. On the otherhand we try to do this refactoring as part when we introduce consumer groups, which would hopefully lessen the use of consumers. But if there are customers using millions of consumers and then having say rate-limiting plugin attached to each consumer, then this might get problematic.

**Comment by hbagdi** [comment]
We shouldn't be differentiating the cardinality of consumers across hybrid mode and traditional mode. I do get that there are technical reasons why different cardinality assumptions exists and how the iterator is built in both modes.

We should keep on aligning both modes as much as possible and try to get to a point where hybrid performs at par with traditional mode. This will take time.

**Comment by bungle** [comment]
> We shouldn't be differentiating the cardinality of consumers across hybrid mode and traditional mode. I do get that there are technical reasons why different cardinality assumptions exists and how the iterator is built in both modes.
> 
> We should keep on aligning both modes as much as possible and try to get to a point where hybrid performs at par with traditional mode. This will take time.

@hbagdi, hybrid/dbless mode is arguable more performant and simpler with normal number of plugins. Iterator just build everything in memory on all the workers. No runtime cache lookups, no runtime db queries when things are not in cache. But with huge number of plugins the change here can cause major issues on those that try to upgrade (aka cannot upgrade, e.g. memory usage jumps from 400 MB to gigabytes per worker or something). As said, I have proposed it before in 2019, which was not given green light, so my question is, is it ok now?

File: kong/runloop/plugins_iterator.lua [hunk 1]
```diff
@@ -539,102 +427,31 @@ function PluginsIterator.new(version)
...
+      local service_id = plugin.service and plugin.service.id
+      local consumer_id = plugin.consumer and plugin.consumer.id
+
+      -- Get the plugin configuration for the specified workspace (ws_id)
+      local cfg = get_plugin_config(plugin, name, ws_id)
+      -- Determine if the plugin configuration is global (i.e., not tied to any route, service, or consumer)
+      local is_global = not route_id and not service_id and not consumer_id
+      if is_global then
+        -- Store the global configuration for the plugin in the 'globals' table
+        globals[name] = cfg
```
Comment (samugi) [inline kong/runloop/plugins_iterator.lua:441]: Is `globals[name]` redundant? I.e. would it always have the same content as `combos[name]["::"]`? `globals` is exposed outside of this module though so maybe we are intentionally keeping the tables separate.

File: spec/02-integration/15-plugins-iterator/02-correctness_spec.lua [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        local r = proxy_client():get("/anything/service-" .. i, {
+          headers = {
+            host = "route.bar." .. i,
+            -- authenticate as `alice`
+            apikey = "alice",
+          },
+        })
+        assert.response(r).has.status(200)
+        -- The plugin for service+1 is not executed
+        assert.request(r).has_no_header("correlation-id-added-service" .. i + 1)
```
Comment (samugi) [inline spec/02-integration/15-plugins-iterator/02-correctness_spec.lua:107]: would it make sense to check *all* the other services (to confirm the plugins did not execute) rather than just the ones immediately before and after?

File: spec/03-plugins/29-acme/05-redis_storage_spec.lua [hunk 3]
```diff
@@ -256,7 +256,7 @@ describe("Plugin: acme (storage.redis)", function()
     end)
   end)
 
-  for _, strategy in helpers.each_strategy() do
+  for _, strategy in helpers.each_strategy({"postgres", "off"}) do
```
Comment (samugi) [inline spec/03-plugins/29-acme/05-redis_storage_spec.lua:259]: it's not obvious to me why this change was needed

Comment (jschmid1) [reply spec/03-plugins/29-acme/05-redis_storage_spec.lua:259]: There's flakyness triggered due to C*'s eventual consistency model. There's probably a more reliable way to ensure consistency of data but I figured we move away from Cassandra anyways and I didn't bother..

File: kong/runloop/plugins_iterator.lua (hunk 1 above)
Comment (jschmid1) [reply kong/runloop/plugins_iterator.lua:441]: I kept the behavior close to the original implementation before we _fully_ understand the behavior of `global` plugins. See https://github.com/Kong/kong-ee/pull/5168

**Review by samugi — APPROVED** [review]

**Review by bungle — APPROVED** [review]
This looks ok. @jschmid1 while you are at it, could you remove the:
https://github.com/Kong/kong/blob/master/kong/runloop/plugins_iterator.lua#L655

This is removed on EE and I don't want us to accidentally reintroduce it back. The exit-transformer runs with Pongo or at least used to, and thus we only catch these errors later. So it is better to remove that line (we have made the error already at least once). This is basically a reason for it: https://github.com/Kong/kong-ee/blob/master/plugins-ee/exit-transformer/kong/plugins/exit-transformer/handler.lua#L29

The discussed global change we can do / try later, if you don't want to try it here.

I am approving this, but please remove that line in this PR, so that we don't make mistake in EE merge.

**Comment by outsinre** [comment]
> @jschmid1 please cherry-pick in EE

It is reverted at https://github.com/Kong/kong/pull/10751. Kick out of `3.3`.
